src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -505,6 +533,33 @@ RouterSequentialModifyExecScan(CustomScanState *node)
 }
 
 
+/*
+ * TaskListRequires2PC determines whether the given task list requires 2PC
+ * because the tasks provided operates on a reference table.
+ *
+ * Note that we currently do not generate tasks lists that involves multiple different
+ * tables, thus we only check the first task in the list.
+ */
+bool
+TaskListRequires2PC(List *taskList)
+{
+	Task *task = NULL;
+
+	if (taskList == NIL)
+	{
+		return false;
+	}
+
+	task = (Task *) linitial(taskList);
+	if (task->replicationModel == REPLICATION_MODEL_2PC)
+	{
+		return true;
```

Comment:
Maybe `if (task->replicationModel != REPLICATION_MODEL_2PC) return false` for consistency.